substrate/client/rpc-spec-v2/src/archive/archive_storage.rs

```diff
@@ -127,3 +139,724 @@ where
+		}
+	}
+
+	/// Check if the key starts with any of the provided items.
+	///
+	/// Returns a `FetchStorage` to indicate if the key should be fetched.
+	fn starts_with(key: &StorageKey, items: &[DiffDetails]) -> Option<FetchStorageType> {
+		// User has requested all keys, by default this fallbacks to fetching the value.
+		if items.is_empty() {
+			return Some(FetchStorageType::Value)
+		}
+
+		let mut value = false;
+		let mut hash = false;
+
+		for item in items {
+			if key.as_ref().starts_with(&item.key.as_ref()) {
+				match item.return_type {
+					ArchiveStorageDiffType::Value => value = true,
+					ArchiveStorageDiffType::Hash => hash = true,
+				}
+			}
+		}
+
+		match (value, hash) {
+			(true, true) => Some(FetchStorageType::Both),
+			(true, false) => Some(FetchStorageType::Value),
+			(false, true) => Some(FetchStorageType::Hash),
+			(false, false) => None,
+		}
+	}
+
+	/// Send the provided result to the `tx` sender.
+	///
+	/// Returns `false` if the sender has been closed.
+	fn send_result(
+		tx: &mpsc::Sender<ArchiveStorageDiffEvent>,
+		result: (StorageResult, Option<StorageResult>),
+		operation_type: ArchiveStorageDiffOperationType,
+		child_trie_key: Option<String>,
+	) -> bool {
+		let res = ArchiveStorageDiffEvent::StorageDiff(ArchiveStorageDiffResult {
+			key: result.0.key,
+			result: result.0.result,
+			operation_type,
+			child_trie_key: child_trie_key.clone(),
+		});
+		if tx.blocking_send(res).is_err() {
+			return false
+		}
+
+		if let Some(second) = result.1 {
+			let res = ArchiveStorageDiffEvent::StorageDiff(ArchiveStorageDiffResult {
+				key: second.key,
+				result: second.result,
+				operation_type,
+				child_trie_key,
+			});
+			if tx.blocking_send(res).is_err() {
+				return false
+			}
+		}
+
+		true
+	}
+
+	fn handle_trie_queries_inner(
+		&self,
+		hash: Block::Hash,
+		previous_hash: Block::Hash,
+		items: Vec<DiffDetails>,
+		tx: &mpsc::Sender<ArchiveStorageDiffEvent>,
+	) -> Result<(), String> {
+		// Parse the child trie key as `ChildInfo` and `String`.
+		let maybe_child_trie = items.first().and_then(|item| item.child_trie_key.clone());
+		let maybe_child_trie_str =
+			items.first().and_then(|item| item.child_trie_key_string.clone());
+
+		// Iterator over the current block.
+		let keys_iter = self.client.raw_keys_iter(hash, maybe_child_trie.clone())?;
+
+		// Iterator over the previous block.
+		//
+		// The hashmap contains the keys and a boolean to indicate if the key is present in the
+		// current hash. This information is later used to determine if the key has been deleted.
+		//
+		// Note: `Itertools::contains` consumes the iterator until the given key is found,
+		// therefore we may lose keys and iterate over the previous block twice for deleted keys.
+		//
+		// Example:
+		// keys_iter: [0, 1, 2]
+		// previous_keys_iter: [1, 2, 3]
+		//  -> the `previous_keys_iter` is entirely consumed while searching for `0`.
+		let mut previous_keys_iter: indexmap::IndexMap<_, _> = self
```

Comment:
I wonder we could just collect two HashSet's one for the previous block and the current then just call [HashSet::difference](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.difference)?